### PR TITLE
[DFSM] Add SHARED_STORAGE_UPDATE_POLICY

### PR DESCRIPTION
**Pull request by chenwany**
### Description of changes
* Add new update policy `SHARED_STORAGE_UPDATE_POLICY` for SharedStorages section, it applies to adding and removing share storages and changing MountDir attribute.
  * It requires compute fleet stop when removing shared storages or changing `MountDir`
  * Update is UNSUPPORT for awsbatch scheduler
  * Computefleet stop is required for plugin scheduler
  * For adding new shared storages, you can either stop compute fleet or apply queue update strategy.

### Tests
* Unit tests
* Manual tests
  * Not stop compute fleet and add EBS, EFS, FSX without queue strategy
  ```
  "message": "All compute nodes must be stopped or QueueUpdateStrategy must be set. Stop the compute fleet with the 
  pcluster update-compute-fleet command, or set QueueUpdateStrategy in the configuration used for the 'update-cluster' 
  operation",
  "currentValue": "-"
  ```

  * Stop and add EBS, EFS, FSX without queue strategy -- OK
  * Remove EBS, EFS or FSX with queue strategy -- All compute nodes must be stopped. Stop the compute fleet with the pcluster update-compute-fleet command
  * Not stop and add EBS, EFS, FSX without queue strategy -- All compute nodes must be stopped or QueueUpdateStrategy must be set. Stop the compute fleet with the pcluster update-compute-fleet command, or set QueueUpdateStrategy in the configuration used for the 'update-cluster' operation"
  * Remove EFS wit queue strategy without stop -- "All compute nodes must be stopped. Stop the compute fleet with the pcluster update-compute-fleet command."
*  For awsbatch, changing the parameters is not support
*  For plugin scheduler, it requires compute fleet stop to add, remove sections or changing MountDir
### Checklist
- [ ] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [ ] Check all commits' messages are clear, describing what and why vs how.
- [ ] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [ ] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,6 +3,8 @@ CHANGELOG
 
 3.x.x
 -----
+**ENHANCEMENTS**
+- Add support for updating SharedStorage configuration.
```
Comment (gmarciani) [inline CHANGELOG.md]: I suggest to highlight the config name as we did in other lines, i.e. `SharedStorage`

File: cli/src/pcluster/config/update_policy.py [hunk 2]
```diff
@@ -125,6 +125,39 @@ def is_slurm_queues_change(change):
     return any(path.startswith("SlurmQueues[") for path in change.path)
 
 
+def is_slurm_scheduler(patch):
+    return patch.cluster.stack.scheduler == "slurm"
```
Comment (gmarciani) [inline cli/src/pcluster/config/update_policy.py]: Here and in other places in our codebase, we redefine the string "slurm".
I think we should define it once in a constant and reference the constant instead.

What about adding such a constant in `cli/src/pcluster/constants.py`?

We can use such constant here and do the larger refactoring later on in a dedicated activity.

File: cli/src/pcluster/config/update_policy.py [hunk 3]
```diff
@@ -125,6 +125,39 @@ def is_slurm_queues_change(change):
...
+
+
+def is_awsbatch_scheduler(_, patch):
+    return patch.cluster.stack.scheduler == "awsbatch"
```
Comment (gmarciani) [inline cli/src/pcluster/config/update_policy.py]: Here and in other places in our codebase, we redefine the string "awsbatch".
I think we should define it once in a constant and reference the constant instead.

What about adding such a constant in cli/src/pcluster/constants.py?

We can use such constant here and do the larger refactoring later on in a dedicated activity.

File: cli/src/pcluster/config/update_policy.py [hunk 4]
```diff
@@ -125,6 +125,39 @@ def is_slurm_queues_change(change):
...
+        return True
+    return False
+
+
+def fail_reason_shared_storage_update_policy(change, patch):
+    if is_awsbatch_scheduler(change, patch):
+        return f"Update actions are not currently supported for the '{change.key}' parameter"
+    reason = "All compute nodes must be stopped"
+    # QueueUpdateStrategy can override UpdatePolicy of parameters under SlurmQueues
+    if not is_stop_required_for_shared_storage(change) and is_slurm_scheduler(patch):
```
Comment (gmarciani) [inline cli/src/pcluster/config/update_policy.py:155]: With these conditions we would have the following case. If the scheduler is Slurm and a stop is not required, we would fail with message "All compute nodes must be stopped or QueueUpdateStrategy must be set". This may confuse the end-user. What do you think?

File: cli/src/pcluster/config/update_policy.py [hunk 5]
```diff
@@ -149,11 +182,47 @@ def condition_checker_queue_update_strategy(change, patch):
...
+    if is_awsbatch_scheduler(change, patch):
+        return (
+            f"Restore '{change.key}' value to '{change.old_value}'"
+            if change.old_value is not None
+            else "{0} the parameter '{1}'".format("Restore" if change.is_list else "Remove", change.key)
+            + ". If you need this change, please consider creating a new cluster instead of updating the existing one."
+        )
+    actions = "Stop the compute fleet with the pcluster update-compute-fleet command"
+    if not is_stop_required_for_shared_storage(change) and is_slurm_scheduler(patch):
+        actions += ", or set QueueUpdateStrategy in the configuration used for the 'update-cluster' operation"
```
Comment (gmarciani) [inline cli/src/pcluster/config/update_policy.py:195]: Missing full stop at the end of the message.

File: cli/src/pcluster/config/update_policy.py [hunk 6]
```diff
@@ -149,11 +182,47 @@ def condition_checker_queue_update_strategy(change, patch):
...
 
+def actions_needed_shared_storage_update(change, patch):
+    if is_awsbatch_scheduler(change, patch):
+        return (
+            f"Restore '{change.key}' value to '{change.old_value}'"
+            if change.old_value is not None
+            else "{0} the parameter '{1}'".format("Restore" if change.is_list else "Remove", change.key)
+            + ". If you need this change, please consider creating a new cluster instead of updating the existing one."
+        )
+    actions = "Stop the compute fleet with the pcluster update-compute-fleet command"
```
Comment (gmarciani) [inline cli/src/pcluster/config/update_policy.py:193]: Missing period at the end of the message if the condition below does not hold.

File: cli/src/pcluster/config/update_policy.py [hunk 7]
```diff
@@ -149,11 +182,47 @@ def condition_checker_queue_update_strategy(change, patch):
...
+        actions += ", or set QueueUpdateStrategy in the configuration used for the 'update-cluster' operation"
+
+    return actions
+
+
+def condition_checker_shared_storage_update_policy(change, patch):
+    """
+    Check different requirements for different schedulers.
+
+    Compute fleet stop is required for awsbatch and plugin scheduler.
```
Comment (gmarciani) [inline cli/src/pcluster/config/update_policy.py]: When the scheduler is AWS Batch then the shared storage update should not be support, right?

File: cli/src/pcluster/config/update_policy.py [hunk 8]
```diff
@@ -149,11 +182,47 @@ def condition_checker_queue_update_strategy(change, patch):
...
+    QueueUpdateStrategy can override UpdatePolicy of parameters under SlurmQueues for slurm scheduler.
+    """
+    if is_awsbatch_scheduler(change, patch):
```
Comment (gmarciani) [inline cli/src/pcluster/config/update_policy.py:208]: We cover awsbatch and slurm here, but the function comment also mentions the scheduler plugin.
What about the scheduler plugin?

File: cli/src/pcluster/config/update_policy.py [hunk 9]
```diff
@@ -149,11 +182,47 @@ def condition_checker_queue_update_strategy(change, patch):
...
+        return None
```
Comment (gmarciani) [inline cli/src/pcluster/config/update_policy.py]: I supposed this function would have returned only True or False. What is the meaning of returning None?

File: cli/src/pcluster/templates/cdk_builder_utils.py [hunk 10]
```diff
@@ -491,7 +491,7 @@ def _build_policy(self) -> List[iam.PolicyStatement]:
             ),
             iam.PolicyStatement(
                 sid="Ec2TagsAndVolumes",
-                actions=["ec2:AttachVolume", "ec2:CreateTags"],
+                actions=["ec2:AttachVolume", "ec2:CreateTags", "ec2:DetachVolume"],
```
Comment (gmarciani) [inline cli/src/pcluster/templates/cdk_builder_utils.py:494]: This kind of change should be covered by unit tests.
I known that we do not have an existing unit test to modify, but we should.
May you please keep track of the lack of unit testing the iam policies created by the template?

File: cli/tests/pcluster/config/test_update_policy.py [hunk 11]
```diff
@@ -453,3 +453,357 @@ def test_compute_fleet_stop_fail_reason_and_actions_needed(
...
+    new_value,
+    update_strategy,
+    expected_result,
+    expected_fail_reason,
+    expected_actions_needed,
+    scheduler,
+):
+    cluster = dummy_cluster()
+    cluster_has_running_capacity_mock = mocker.patch.object(
+        cluster, "has_running_capacity", return_value=not is_fleet_stopped
```
Comment (gmarciani) [inline cli/tests/pcluster/config/test_update_policy.py:773]: Minor comment: why having a variable `is_fleet_stopped` to negate instead of having a variable `is_fleet_running` to check positively?

File: cli/tests/pcluster/config/test_update_policy.py [hunk 12]
```diff
@@ -453,3 +453,357 @@ def test_compute_fleet_stop_fail_reason_and_actions_needed(
...
+            [],
+            None,
+            {"MountDir": "/ebs3", "Name": "ebs3", "StorageType": "Ebs", "EbsSettings": {"VolumeType": "gp3"}},
+            None,
+            False,
+            "All compute nodes must be stopped or QueueUpdateStrategy must be set",
+            "Stop the compute fleet with the pcluster update-compute-fleet command, or set QueueUpdateStrategy in the "
+            "configuration used for the 'update-cluster' operation",
+            "slurm",
+            id="running fleet and EBS changed with no update strategy set",
```
Comment (gmarciani) [inline cli/tests/pcluster/config/test_update_policy.py]: We say here that the test case is "EBS changed", but we are actually adding a new EBS section right? If this is the case, let's rather say "add a new EBS section".

File: cli/tests/pcluster/config/test_update_policy.py [hunk 13]
```diff
@@ -453,3 +453,357 @@ def test_compute_fleet_stop_fail_reason_and_actions_needed(
...
+            [],
+            None,
+            {"MountDir": "/ebs3", "Name": "ebs3", "StorageType": "Ebs", "EbsSettings": {"VolumeType": "gp3"}},
+            None,
+            None,
+            "Update actions are not currently supported for the 'SharedStorage' parameter",
+            "Restore the parameter 'SharedStorage'. If you need this change, please consider creating a new cluster "
+            "instead of updating the existing one.",
+            "awsbatch",
+            id="running fleet and EBS changed with no update strategy set with batch scheduler",
```
Comment (gmarciani) [inline cli/tests/pcluster/config/test_update_policy.py]: Typically we refer to the AWS Batch scheduler with the name "awsbatch" rather than "batch".

File: cli/tests/pcluster/config/test_update_policy.py [hunk 14]
```diff
@@ -453,3 +453,357 @@ def test_compute_fleet_stop_fail_reason_and_actions_needed(
...
+            "SharedStorage",
+            [],
+            None,
+            {"MountDir": "/ebs3", "Name": "ebs3", "StorageType": "Ebs", "EbsSettings": {"VolumeType": "gp3"}},
+            QueueUpdateStrategy.TERMINATE.value,
+            True,
+            None,
+            None,
+            "slurm",
+            id="running fleet and Ebs added with update strategy TERMINATE",
```
Comment (gmarciani) [inline cli/tests/pcluster/config/test_update_policy.py]: In `id` string here and below: Ebs => EBS

File: cli/tests/pcluster/config/test_update_policy.py [hunk 15]
```diff
@@ -453,3 +453,357 @@ def test_compute_fleet_stop_fail_reason_and_actions_needed(
...
+            "SharedStorage",
+            [],
+            None,
+            {"MountDir": "/efs4", "Name": "efs4", "StorageType": "Efs"},
+            QueueUpdateStrategy.TERMINATE.value,
+            True,
+            None,
+            None,
+            "slurm",
+            id="running fleet and EFs added with update strategy DRAIN",
```
Comment (gmarciani) [inline cli/tests/pcluster/config/test_update_policy.py]: The update strategy is actually TERMINATE, not DRAIN.

**Review by gmarciani — CHANGES_REQUESTED** [review]

File: cli/src/pcluster/config/update_policy.py (hunk 2 above)
Comment (chenwany) [reply cli/src/pcluster/config/update_policy.py]: added

File: cli/src/pcluster/config/update_policy.py (hunk 3 above)
Comment (chenwany) [reply cli/src/pcluster/config/update_policy.py]: added to constants.py

File: cli/src/pcluster/config/update_policy.py (hunk 4 above)
Comment (chenwany) [reply cli/src/pcluster/config/update_policy.py:155]: Could you explain why do you think this is confusing? This is aligned with the error message when Parameters require QueueUpdateStrategy.

File: cli/src/pcluster/config/update_policy.py (hunk 5 above)
Comment (chenwany) [reply cli/src/pcluster/config/update_policy.py:195]: This is aligned with the error message when Parameters require QueueUpdateStrategy.

File: cli/src/pcluster/config/update_policy.py (hunk 7 above)
Comment (chenwany) [reply cli/src/pcluster/config/update_policy.py]: Yup, thanks, update the comment

File: cli/src/pcluster/config/update_policy.py (hunk 8 above)
Comment (chenwany) [reply cli/src/pcluster/config/update_policy.py:208]: The function also cover scheduler plugin. line `result = not patch.cluster.has_running_capacity()`, cover scheduler plugin case

File: cli/src/pcluster/config/update_policy.py (hunk 9 above)
Comment (chenwany) [reply cli/src/pcluster/config/update_policy.py]: good catch, this should return False

File: cli/tests/pcluster/config/test_update_policy.py (hunk 11 above)
Comment (chenwany) [reply cli/tests/pcluster/config/test_update_policy.py:773]: I think `is_fleet_stopped` is a name better explain itself. Refactoring of `is_fleet_stopped` is a function defined before and it is out of scope here.
Also, `is_fleet_stopped` is not opposite to `is_fleet_running`. If the compute fleet have 0 compute nodes, the cluster compute fleet status is running, but it is not Is_fleet_stop.

File: cli/tests/pcluster/config/test_update_policy.py (hunk 12 above)
Comment (chenwany) [reply cli/tests/pcluster/config/test_update_policy.py]: Changed the id

File: cli/tests/pcluster/config/test_update_policy.py (hunk 13 above)
Comment (chenwany) [reply cli/tests/pcluster/config/test_update_policy.py]: Changed the id

File: cli/tests/pcluster/config/test_update_policy.py (hunk 14 above)
Comment (chenwany) [reply cli/tests/pcluster/config/test_update_policy.py]: Changed the id

File: cli/tests/pcluster/config/test_update_policy.py (hunk 15 above)
Comment (chenwany) [reply cli/tests/pcluster/config/test_update_policy.py]: Changed the id

File: cli/src/pcluster/templates/cdk_builder_utils.py (hunk 10 above)
Comment (chenwany) [reply cli/src/pcluster/templates/cdk_builder_utils.py:494]: Created a SIM to track it

File: cli/src/pcluster/config/update_policy.py (hunk 4 above)
Comment (gmarciani) [reply cli/src/pcluster/config/update_policy.py:155]: If a stop is not required, but we receive a message saying "All compute nodes must be stopped", then this sounds to be confusing. 
Nevertheless, I agree that we must be consistent with the messaging returned by QueueUpdateStrategy, so let's verify with the beta testing if such messages create confusion or not in the end user.
Let's proceed as it is.

File: cli/src/pcluster/config/update_policy.py (hunk 5 above)
Comment (gmarciani) [reply cli/src/pcluster/config/update_policy.py:195]: In general, I think that messages should end with a period.
Nevertheless, as long as we are consistent with ourselves that's fine. Let's go with it.

File: cli/src/pcluster/config/update_policy.py (hunk 8 above)
Comment (gmarciani) [reply cli/src/pcluster/config/update_policy.py:208]: Ok, got it

File: cli/src/pcluster/config/update_policy.py (hunk 6 above)
Comment (gmarciani) [reply cli/src/pcluster/config/update_policy.py:193]: In general, I think that messages should end with a period.
Nevertheless, as long as we are consistent with ourselves that's fine. Let's go with it.

File: cli/src/pcluster/templates/cdk_builder_utils.py (hunk 10 above)
Comment (gmarciani) [reply cli/src/pcluster/templates/cdk_builder_utils.py:494]: Thank you so much.

File: cli/tests/pcluster/config/test_update_policy.py (hunk 11 above)
Comment (gmarciani) [reply cli/tests/pcluster/config/test_update_policy.py:773]: Got it thanks!

**Review by gmarciani — DISMISSED** [review]

**Review by gmarciani — COMMENTED** [review]
Diffs look good to me. Nevertheless, some checks are failing so please verify and fix the root cause before merging.

**Review by gmarciani — APPROVED** [review]

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 16]
```diff
@@ -469,7 +471,9 @@ class SharedStorageSchema(BaseSchema):
     """Represent the generic SharedStorage schema."""
 
     mount_dir = fields.Str(
-        required=True, validate=get_field_validator("file_path"), metadata={"update_policy": UpdatePolicy.UNSUPPORTED}
+        required=True,
+        validate=get_field_validator("file_path"),
+        metadata={"update_policy": UpdatePolicy.SHARED_STORAGE_UPDATE_POLICY},
```
Comment (hanwen-cluster) [inline cli/src/pcluster/schemas/cluster_schema.py:476]: Why is this change needed? Isn't the only change needed on [line 1942](https://github.com/aws/aws-parallelcluster/pull/4255/files#diff-ea5b87cce964033aecc0072b33dd35596bf4f5a3d63ae4dd76fb7b318ba22318R1942)
